addons/binding/org.openhab.binding.groheondus/src/main/java/org/openhab/binding/groheondus/internal/handler/GroheOndusAccountHandler.java

```diff
@@ -51,6 +66,43 @@ public OndusService getService() {
         return ret;
     }
 
+    public void deleteRefreshToken() {
+        this.storage.remove(STORAGE_KEY_REFRESH_TOKEN);
+        this.initialize();
+
+        if (refreshTokenFuture != null) {
+            refreshTokenFuture.cancel(true);
+        }
+    }
+
+    public void setRefreshToken(String refreshToken) {
+        this.storage.put(STORAGE_KEY_REFRESH_TOKEN, refreshToken);
```

Comment:
Please use the thing configuration and not a separate storage bucket.
Write the token into a field that has the context set to "password" in the xml file.

Reply:
I kind of did that intentionally, as I don't want to require people to put their refresh token into a configuration file (like they needed to do with the username/password). If one configures their things in thing-files, then they can not put the refresh token into the thing using the paper UI, iirc. So they would be forced to put the refresh token in plain text into the configuration file. To prevent this, I thought storing it in a separate storage makes sense. Or do you've another idea how to solve this problem? :)

Reply:
This is a shortage on openHab core, I agree, but should not be sorted by binding developers. openHAB 3 will do that differently, but either not having .thing and .item files altogether or separating runtime and configuration data.

Reply:
This seems to be an OAuth2 authentication. openhab-core has a built client for OAuth2 authentication. This also usages storage and actually encrypts the refreshtoken when storing it. Since the groheondus binding uses an external library that both handles the authentication as well as the api communication it's not trivial to use the OAuth2 client of openHAB, without hooking deep into the grohe library. Ideally you would use openHAB for authentication and the grohe library for calling the api. You might even be able to use the OAuth2 openHAB client only for storing the refreshtoken, but that would be a kind of special usage. If you want to learn more about the openHAB OAuth2 client take a look at the Spotify binding it uses this.

Reply:
@Hilbrand  How do we proceed if he cannot use the core oauth2 library? I guess storing it as plain text like other bindings need to do it at the moment would be the cleanest solution?

Reply:
I don't think, that using the core OAuth2 component is easily useable, as the OAuth2 specific parts are already handled by the grohe library already. So the only thing that needs to be done is storing the access and refresh tokens. Wouldn't it be possible, for this use case, to use the `OAuthStoreHandler` directly? In that way, the storage would be whatever the core component needs to do (e.g. encrypting and decrypting the tokens), but the tokens would still be useable by the grohe binding?